# $\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$ΑΣΚΗΣΗ 3
$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$Βασίλης Πουλος, 2805
$\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;\;$Κωστάκης Ελευθέριος-Παναγιώτης, 2741


### Φτιάχνουμε τον αρχικό πίνακα μια φορά στην αρχή και για τις δυο εκδοχές του αλγορίθμου.

In [2]:
import pandas as pd
import numpy as np

# Creating the database with the ratings
user_ratings = pd.DataFrame({'Movie 1':[5.0, 4.0, 3.0, 2.0], 'Movie 2': [3, 2, None, 5],
                             'Movie 3':[None, 1, None, 1], 'Movie 4': [1, None, 1, 5],
                             'Movie 5': [None, None, 3, 3], 'Movie 6': [None, 1, 3, 4]},
                             index = ['User X', 'User Y', 'User Z', 'User W'])
user_ratings

Movie 1  Movie 2  Movie 3  Movie 4  Movie 5  Movie 6
User X      5.0      3.0      NaN      1.0      NaN      NaN
User Y      4.0      2.0      1.0      NaN      NaN      1.0
User Z      3.0      NaN      NaN      1.0      3.0      3.0
User W      2.0      5.0      1.0      5.0      3.0      4.0

### Υλοποίηση της πρώτης εκδοχής  του αλγορίθμου

In [86]:
from sklearn.metrics.pairwise import cosine_similarity
from scipy import sparse
import heapq

A =  np.array([user_ratings.loc['User X'].values,
               user_ratings.loc['User Y'].values,
               user_ratings.loc['User Z'].values,
               user_ratings.loc['User W'].values])

A_sparse = sparse.csr_matrix(A)
A_sparse = np.nan_to_num(A_sparse.toarray())

similarities = cosine_similarity(A_sparse)

# Print Similarity table for User X
from prettytable import PrettyTable
t = PrettyTable(['Similarities','User X', 'User Y', 'User Z', 'User W'])
t.add_row(['User X', similarities[0][0], similarities[0][1], similarities[0][2], similarities[0][3]])
print(t)

# Get the 2 largest values and its' index value
index = heapq.nlargest(2, range(1,len(similarities[0])), similarities[0].take)
values = [similarities[0][index[0]], similarities[0][index[1]]]
provlepsi = values[0]*user_ratings.iloc[index[0]].values[5] + values[1]*user_ratings.iloc[index[1]].values[5]
provlepsi /= values[0]+values[1]
print('\n'+'\033[1m'+'\033[4m'+'H provlepsi gia tin kritiki (X,6) einai:'+'\033[1m'+'\033[0m', provlepsi)

+--------------+--------+--------------------+--------------------+--------------------+
| Similarities | User X |       User Y       |       User Z       |       User W       |
+--------------+--------+--------------------+--------------------+--------------------+
|    User X    |  1.0   | 0.9369749612033813 | 0.5111012519999519 | 0.5669467095138409 |
+--------------+--------+--------------------+--------------------+--------------------+

H provlepsi gia tin kritiki (X,6) einai: 2.1309366449453377


### Υλοποίηση της δεύτερης εκδοχής  του αλγορίθμου

In [85]:
from sklearn.metrics.pairwise import cosine_similarity
from scipy import sparse
import heapq

mean_ratings_dev = user_ratings.sub(user_ratings.mean(axis = 1, skipna = True), axis = 0)
print('\n'+'\033[1m'+'\033[4m'+'User ratings afairontas tin mesi timh kathe grammhs:'+'\033[1m'+'\033[0m')
print(mean_ratings_dev)

A =  np.array([mean_ratings_dev.loc['User X'].values,
               mean_ratings_dev.loc['User Y'].values,
               mean_ratings_dev.loc['User Z'].values,
               mean_ratings_dev.loc['User W'].values])

A_sparse = sparse.csr_matrix(A)
A_sparse = np.nan_to_num(A_sparse.toarray())

similarities = cosine_similarity(A_sparse)
# Print Similarity table for User X
from prettytable import PrettyTable
t = PrettyTable(['Similarities','User X', 'User Y', 'User Z', 'User W'])
t.add_row(['User X', similarities[0][0], similarities[0][1], similarities[0][2], similarities[0][3]])
print('\n'+'\033[1m'+'\033[4m'+'Similarity table gia ton User X'+'\033[1m'+'\033[0m')
print(t)

index = heapq.nlargest(2, range(1,len(similarities[0])), similarities[0].take)
values = [similarities[0][index[0]], similarities[0][index[1]]]

provlepsi = values[0]*mean_ratings_dev.iloc[index[0]].values[5] + values[1]*mean_ratings_dev.iloc[index[1]].values[5]
provlepsi /= values[0]+values[1]
provlepsi += user_ratings.loc['User X'].mean(skipna = True)
print('\n'+'\033[1m'+'\033[4m'+'H provlepsi gia tin kritiki (X,6) einai:'+'\033[1m'+'\033[0m',provlepsi)


User ratings afairontas tin mesi timh kathe grammhs:
         Movie 1   Movie 2   Movie 3   Movie 4   Movie 5   Movie 6
User X  2.000000  0.000000       NaN -2.000000       NaN       NaN
User Y  2.000000  0.000000 -1.000000       NaN       NaN -1.000000
User Z  0.500000       NaN       NaN -1.500000  0.500000  0.500000
User W -1.333333  1.666667 -2.333333  1.666667 -0.333333  0.666667

Similarity table gia ton User X
+--------------+--------------------+--------------------+-------------------+---------------------+
| Similarities |       User X       |       User Y       |       User Z      |        User W       |
+--------------+--------------------+--------------------+-------------------+---------------------+
|    User X    | 0.9999999999999998 | 0.5773502691896258 | 0.816496580927726 | -0.5809475019311123 |
+--------------+--------------------+--------------------+-------------------+---------------------+

H provlepsi gia tin kritiki (X,6) einai: 2.878679656440357


<br><br><br>
## Ανάλυση των αλγορίθμων και αποτελεσμάτων


$\;\;\;$Παρατηρούμε ότι υπάρχει αξιοσημείωτη διαφορά στις δύο περιπτώσεις, αυτό συμβαίνει διότι στην πρώτη εκδοχή του αλγορίθμου υπολογίζουμε την εκτίμιση (Χ,6) βάσει των βαθμολογιών των χρηστών, ενώ στην δεύτερη παίρνουμε την απόκλιση από την μέση τιμή των βαθμολογιών για κάθε χρήστη.<br><br>
$\;\;\;$Έτσι σε κάθε περίπτωση παίρνουμε διαφορετικά cosine similarities.
Στην πρώτη εκδοχή οι χρήστες με την μεγαλύτερη ομοιότητα είναι ο Υ με cos_sim(X,Y)=0,93 και ο W με cos_sim(X,W)=0.56 ενώ στην δεύτερη εκδοχή έχουμε τον Ζ με τον Υ με cos_sim(X,Z)=0.81 και cos(X,Y)=0.57 αντίστοιχα.


$\;\;\;$Αυτές οι αλλαγές υπάρχουν διότι κλαθε εκδοχή 'δουλεύει' διαφορετικά. Συγκεκριμένα για την πρώτη εκδοχή υπολογίζεται η ομοιότητα βάσει τις τιμές των βαθμολογιών, ενώ στην δεύτερη παίζει ρόλο η ομοιότητα ανάμεσα στην τυπική απόκλιση των βαθμολογιών. 

$\;\;\;$Στην πρώτη οι χρήστες με τη μεγαλύτερη βαρύτητα ως προς το αποτέλεσμα είναι ο Y=1 και ο W=4 με τον Υ να έχει μεγαλύτερο βάρος cos_sim(X,Y)=0.936 από τον W με cos_sim(X,W)=0,56. Στην περίπτωση που κοιτάμε τον Y σε σχέση με τον X, έχει μεγάλη ομοιότητα διότι στις ίδιες ταινίες έχουν δώσει παρόμοιες βαθμολογίες, για την 'Movie 1' ο Χ και Y έχουν βαθμολογίσει με 5 και 4 αντίστοιχα καθώς και στην ταινία 2 έχουν δώσει 3 και 2 αντίστοιχα, έτσι αυξάνεται σημαντικά ο αριθμιτής στον τύπο της ομοιότητας. Ενώ κάτι τέτοιο δεν ισχυέι για τον χρήστη W σε σχέση με τον Χ, όπου έχουν βαθμολογίσει τις ίδιες ταινίες με αρκετά διαφορετική βαθμολογία. Ο Χ δίνει 5 και 3 όπως είπαμε για 'Movie 1' και 'Movie 2' ενώ ο W με 2 και 5 αντίστοιχα. Στην ταινία 6 ο Y δίνει 1 και ο W 4, οπότε διαισθητικά λέμε ότι η βαθμολογία θα είναι σαφώς πιο κοντά σε αυτήν που έδωσε ο Υ παρά ο W. Και έτσι είναι με 2.13.<br>

$\;\;\;$Στην δεύτερη περίπτωση κοιτάμε διαφορετικούς χρήστες. Την μεγαλύτερη ομοιότητα με τον Χ έχουν οι χρήστες Ζ με τιμή 0.816 και Υ με 0.577. Εδώ ο Ζ επηρεάζει περισσότερο τον Χ και δίνει βαθμολογία 0.5(με τον νέο πίνακα) για την ταινία 6 με τις ταινίες που παίζουν ρόλο να είναι οι 'Movie 1', 'Movie 5' και 'Movie 6' γιατί δίνουν πολύ μικρή τυπική απόκλιση, όντας πιο σταθερός στις επιλογές του και έτσι να υπάρχει μεγαλύτερη ομοιότητα. Ο 'Υ' δίνει -1. Επομένως η βαθμολογία θα πρέπει να βρίσκεται περισσότερο κοντά σε αυτήν του χρήστη Ζ και σε συνδιασμό με την αρνητική βαθμολογία που δίνει ο Υ η προβλεπόμενη κριτική είναι 2.87.